brouter-core/src/main/java/btools/router/StdPath.java

```diff
@@ -139,7 +145,31 @@ else if ( ehbu < 0 )
     }
 
     sectionCost += dist * costfactor + 0.5f;
-      
+
+    if (rc.bikeMode || rc.footMode) {
+      // Uphill angle
+      double alpha = Math.atan2(delta_h, distance);
+
+      // Travel speed
+      double speed = Double.NaN;
+      if (rc.footMode) { // TODO: || tags['way'].search('bicycle=dismount') !== -1) {
+        // Use Tobler's hiking function for walking sections
+        speed = 6 * Math.exp(-3.5 * Math.abs(delta_h / distance + 0.05)) / 3.6;
+      } else {
+        // Compute the speed assuming a basic kinematic model with constant
+        // power.
+        Cubic speedEquation = new Cubic();
+        speedEquation.solve(rc.S_C_x, 0.0, (rc.bikeMass * GRAVITY * (rc.defaultC_r + Math.sin(alpha))), -1.0 * rc.bikerPower);
```

Comment:
I'm still not very well versed into the BRouter routing, so I'm not sure whether this is done for *every* segment which is even considered or only for the segments which will be part of the final track.

If that's the latter case, I guess this is rather performance-sensitive and solving the Cubic equation there (although the code could be optimized a bit) might not be the best idea :/

Reply:
Hi Lucas,

Yes, StdPath.processWaySection is called for every segment evaluated by thr router. However, in the calling class OsmPath there's a flag "detailMode", and if you make that available in processWaySection  you can use that to do this calculation only for the final pass, which touches far less way sections.

Guess you could replace the cubic formula by newton approximation. Just need a good starting value.

One more remark: the whole idea I think is to catch the fact that going up and down is slower on average than going straight. But to achieve that you should use the filtered elevation-delta, because other wise the elevation noise made by buildings will have the same effect than real hills.

regards, Arndt